rewrite-core/src/main/java/org/openrewrite/internal/ListUtils.java

```diff
@@ -127,11 +128,15 @@ public static <T> List<T> map(@Nullable List<T> ls, BiFunction<Integer, T, T> ma
                 }
                 newLs.set(i, newTree);
             }
+            nullEncountered |= newTree == null;
         }
 
-        if (newLs != ls) {
-            //noinspection StatementWithEmptyBody
-            while (newLs.remove(null)) ;
+        if (newLs != ls && nullEncountered) {
+            for (int i = newLs.size() - 1; i >= 0; i--) {
+                if (newLs.get(i) == null) {
+                    newLs.remove(i);
+                }
+            }
```

Comment:
This change isn't related to `flatMap()`, but it uses the same pattern. While it is slightly more verbose it has the advantage that the list is at most traversed once, whereas previously the list would be traversed once for every `null` element. If there are no `null`s the list is not traversed at all, which I think is nice because that will typically be the case.